runtime/src/iree/hal/drivers/hip/native_executable.c

```diff
@@ -252,136 +263,180 @@ iree_status_t iree_hal_hip_native_executable_create(
+                                                                 executable
+                                                                     ->per_device_data
+                                                                         [0])) +
+                                                            (i *
+                                                             native_executable_device_info_size));
+  }
+
   iree_status_t status = iree_ok_status();
-  for (iree_host_size_t i = 0; i < module_count; ++i) {
-    iree_hal_hip_ModuleDef_table_t module_def =
-        iree_hal_hip_ModuleDef_vec_at(modules_vec, i);
 
-    // WARNING: HIP doesn't take an expected length here so we can't bound it.
-    // It's likely that users could craft inputs that read beyond the extents of
-    // the embedded binary.
-    flatbuffers_string_t hsaco_image =
-        iree_hal_hip_ModuleDef_hsaco_image_get(module_def);
-
-    // TODO: pass hipJitOption values to get log info and other info back.
-    // We pass the error buffer today but could use the info log to diagnose
-    // performance warnings.
-    char error_log[8192] = {0};
-    hipJitOption jit_options[] = {
-        hipJitOptionErrorLogBuffer,
-        hipJitOptionErrorLogBufferSizeBytes,
-    };
-    void* jit_option_values[] = {
-        (void*)error_log,
-        (void*)(uint32_t)sizeof(error_log),
-    };
-    hipModule_t module = NULL;
+  for (size_t j = 0; j < num_devices && iree_status_is_ok(status); ++j) {
     status = IREE_HIP_RESULT_TO_STATUS(
-        symbols,
-        hipModuleLoadDataEx(&module, hsaco_image, IREE_ARRAYSIZE(jit_options),
-                            jit_options, jit_option_values),
-        "hipModuleLoadDataEx");
-    if (!iree_status_is_ok(status)) {
-      status = iree_status_annotate(
-          status,
-          IREE_SV("mismatched target chip? missing/wrong bitcode directory?"));
-      if (strlen(error_log) > 0) {
-        status =
-            iree_status_annotate(status, iree_make_cstring_view(error_log));
-      }
+        symbols, hipCtxPushCurrent(devices[j].hip_context),
+        "hipCtxPushCurrent");
+
+    if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
       break;
     }
-
-    executable->modules[i] = module;
-  }
-
-  if (iree_status_is_ok(status)) {
-    for (iree_host_size_t i = 0; i < export_count; ++i) {
-      iree_hal_hip_ExportDef_table_t export_def =
-          iree_hal_hip_ExportDef_vec_at(exports_vec, i);
-
-      // Lookup the function in the module; this should always succeed but
-      // we cannot trust that the input was generated by our compiler.
-      uint32_t module_ordinal =
-          iree_hal_hip_ExportDef_module_ordinal_get(export_def);
-      hipModule_t module = executable->modules[module_ordinal];
-      flatbuffers_string_t kernel_name =
-          iree_hal_hip_ExportDef_kernel_name_get(export_def);
-      hipFunction_t function = NULL;
+    iree_hal_hip_native_executable_per_device_data_t* ned =
```

Comment:
![image](https://github.com/user-attachments/assets/717ee2ab-be88-4ed8-bf95-de72ce961371)

```suggestion
    iree_hal_hip_native_executable_per_device_data_t* per_device_data =
```